packages/core/src/factory.ts

```diff
@@ -33,101 +35,124 @@ export function factory<Resource, Input, Output, Config, LoadOpts>(
         }
       },
 
-      process: async (input: Input, c?: Config) => {
+      process: (input: Input, c?: Config) => {
         // build the config for this request
         const configMerger = configMergerFactory(defaultConfig, customConfig, c);
         const configObj: Config | undefined = configMerger(input);
         const inputValidations: IPrismDiagnostic[] = [];
 
-        // find the correct resource
-        let resource: Resource | undefined;
         if (components.router) {
-          try {
-            resource = components.router.route({ resources, input, config: configObj }, defaultComponents.router);
-          } catch (error) {
-            // rethrow error we if we're attempting to mock
-            if ((configObj as IPrismConfig).mock) {
-              throw error;
-            }
-            const { message, name, status } = error as ProblemJsonError;
-            // otherwise let's just stack it on the inputValidations
-            // when someone simply wants to hit an URL, don't block them
-            inputValidations.push({
-              message,
-              source: name,
-              code: status,
-              severity: DiagnosticSeverity.Warning,
-            });
-          }
-        }
+          return components.router
+            .route({ resources, input, config: configObj }, defaultComponents.router)
+            .fold(
+              error => {
+                // rethrow error we if we're attempting to mock
+                if ((configObj as IPrismConfig).mock) {
+                  return left2v(error);
+                }
 
-        // validate input
-        if (resource && components.validator && components.validator.validateInput) {
-          inputValidations.push(
-            ...(await components.validator.validateInput(
-              {
-                resource,
-                input,
-                config: configObj,
-              },
-              defaultComponents.validator,
-            )),
-          );
-        }
+                const { message, name, status } = error as ProblemJsonError;
+                // otherwise let's just stack it on the inputValidations
+                // when someone simply wants to hit an URL, don't block them
+                inputValidations.push({
```

Comment:
Pushing things to array which is defined outside function chains feels not right. How about using validations? (https://dev.to/gcanti/getting-started-with-fp-ts-either-vs-validation-5eja - Validation section in the middle)

Reply:
Yes! That's exactly what I wanted to do. However I'm proceeding with baby steps here as — you've probably noticed — it's hard to digest all in one place; moreover we're close to the release and I didn't want to change that many things at the time.

…but yes it's exactly the direction I want to take!